numpy/core/src/umath/ufunc_object.c

```diff
@@ -2299,19 +2387,104 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
      * (Just checks core; broadcast dimensions are tested by the iterator.)
      */
     for (i = 0; i < nop; i++) {
-        if (op[i] != NULL && PyArray_NDIM(op[i]) < core_num_dims[i]) {
-            PyErr_Format(PyExc_ValueError,
+        /* TODO: store _n_required after parsing, ahead-of-time */
+        int n_required = ufunc->core_num_dims[i];
+        if (n_required > 0) {
+            /* can be less if dim is flexible */
+            for (j = ufunc->core_offsets[i];
```

Comment:
Predefine `core_offset` and `num_dims` at the start of the loop to make this and the loops below a bit easier to parse.

Reply:
fixed